lib/entry.js

```diff
@@ -0,0 +1,180 @@
+  <Switch>
+    {render()}
+    <Route render={props => (
+      <Index
+        {...props}
+        routes={routes}
+      />
+    )} />
+  </Switch>
+)
+
+class Catch extends React.Component {
+  static getDerivedStateFromProps (props, state) {
+    if (!state.err) return null
+    return { err: null }
+  }
+
+  state = {
+    err: null
+  }
+
+  componentDidCatch (err) {
+    this.setState({ err })
+  }
+
+  render () {
+    const { err } = this.state
+
+    if (err) {
+      return (
+        <pre
+          children={err.toString()}
+          style={{
+            color: 'white',
+            backgroundColor: 'red',
+            fontFamily: 'Menlo, monospace',
+            fontSize: '14px',
+            margin: 0,
+            padding: '16px',
+            minHeight: '128px',
+            whiteSpace: 'prewrap'
+          }}
+        />
+      )
+    }
+
+    return this.props.children
+  }
+}
+
+const Router = IS_CLIENT ? BrowserRouter : StaticRouter
+const App = withRouter(APP ? (require(APP).default || require(APP)) : DefaultApp)
+
+export const getRoutes = async (components = initialComponents) => {
+  const routes = await components.map(async ({ key, name, Component }) => {
+    const exact = name === index
+    let pathname = exact ? '/' : '/' + name
+    const props = Component.getInitialProps
+      ? await Component.getInitialProps({ path: pathname })
+      : {}
+    pathname = props.path || pathname
+    return {
+      key: name,
+      name,
+      path: pathname,
+      exact,
+      Component,
+      props
+    }
+  })
+  return Promise.all(routes)
+}
+
+export default class Root extends React.Component {
+  static defaultProps = {
+    path: '/',
+    basename
+  }
+  state = this.props
+
+  render () {
+    const {
+      routes,
+      basename,
+      path = '/'
+    } = this.state
+
+    return (
+      <Router
+        context={{}}
+        basename={basename}
+        location={path}>
+        <React.Fragment>
+          <Catch>
+            <App
+              routes={routes}
+              render={appProps => (
+                routes.map(({ Component, ...route }) => (
+                  <Route
+                    {...route}
+                    render={props => (
+                      <Catch>
+                        <Component
+                          {...props}
+                          {...appProps}
+                          {...route.props}
+                        />
+                      </Catch>
+                    )}
+                  />
+                ))
+              )}
+            />
+          </Catch>
+        </React.Fragment>
+      </Router>
+    )
+  }
+}
+
+let app
+if (IS_CLIENT) {
+  const mount = DEV ? render : hydrate
+  const div = window.root || document.body.appendChild(
+    document.createElement('div')
+  )
+  getRoutes()
+    .then(routes => {
+      app = mount(<Root routes={routes} />, div)
+    })
+}
+
+if (IS_CLIENT && module.hot) {
+  module.hot.accept()
```

Comment:
*Apparently* this is all thats needed for HMR, but probably need to monkey test this